[Sema] Perform name lookup in outer scopes in some cases involving conditional conformances

huonw opened the pull request:
~Work towards a possible solution for rdar://problem/37760678. Seeing how bad the performance impact might be.~ We decided this approach may be okay.

huonw committed:
[AST] Use flags instead of booleans for UnqualifiedLookup.

huonw committed:
[AST] Allow unqualified name lookup to return results from outer scopes.

This is necessary for disambiguating calls to functions like `min` and
`max`.

huonw commented:
@swift-ci please smoke test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw committed:
[Sema] Store the outer results of lookup for unqualified possible-method-calls.

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci please test compiler performance

xedin commented on include/swift/AST/NameLookup.h:
```diff
@@ -86,19 +86,37 @@ struct LookupResultEntry {
 /// unqualified lookup (i.e. lookup for a plain identifier).
 class UnqualifiedLookup {
 public:
+  enum class Flags {
```
This seems like a nice improvement over boolean flags by itself!

huonw commented:
@swift-ci please test compiler performance

huonw commented:
@swift-ci Please test compiler performance

huonw committed:
[CS] Add decl-ref outer alternatives as non-favored options.

Given something like `max(1, 2)` inside a type that conforms to Sequence, this
allows the compiler to consider Swift.max as well as the two methods with that
name on Sequence.

huonw committed:
[Sema] Restrict outer-lookup to 'min' and 'max' from conditional conformances.

This is a semantics-changing change, and we want it to minimize a source
break (rdar://problem/37760678), but imposing this behaviour by default for
everything is not appropriate with more design and a S-E proposal.

Fixes rdar://problem/37760678

huonw commented:
I've now got this working, restricted to name lookup for `min` and `max` (I'll try `type` tomorrow) where those `min` and `max` come from a conditional conformance, to focus down this semantics-changing change to as limited scope as possible (it's seems like doing it properly/for everywhere would at least require some sort of core team discussion if not a full S-E).

I'm considering even saying that this is deprecated name-lookup: that is, include a warning (with fixit for adding the appropriate `module.` prefix) when one of the "incorrect"/outer functions is chosen, so that we can restrict this source-compat hack behaviour to Swift 4. (I guess Slava will say that we should just decide to do full stack name lookup/constraint solving for everything...)

jrose-apple commented:
I like the deprecation approach. We did something similar for `print` back when that was causing issues.

DougGregor commented on lib/Sema/TypeCheckNameLookup.cpp:
```diff
@@ -26,10 +26,23 @@
 
-void LookupResult::filter(const std::function<bool(LookupResultEntry)> &pred) {
+void LookupResult::filter(
+    const std::function<bool(LookupResultEntry, bool)> &pred) {
```
Huh, we should have used `llvm::function_ref` here.

DougGregor commented on test/NameBinding/name_lookup_min_max_conditional_conformance.swift:
```diff
@@ -0,0 +1,50 @@
+    func f() {
+        _ = max(1, 2)
+        // expected-error@-1{{use of 'max' refers to instance method}}
+        // expected-note@-2{{use 'Swift.' to reference the global function}}
```
Diagnostics here are great, thank you.

DougGregor commented on lib/Sema/TypeCheckConstraints.cpp:
```diff
@@ -397,6 +397,51 @@ static bool diagnoseOperatorJuxtaposition(UnresolvedDeclRefExpr *UDRE,
+  return AllDeclRefs;
+}
+
+static bool shouldConsiderOuterResultsFor(DeclName name) {
```
Could you add a comment here about the list being very restrictive because this is a source-compatibility issue?

DougGregor reviewed (approved):
That core logic in `UnqualifiedLookup` is pretty hairy, but I've convinced myself it's right. Thank you!

huonw committed:
[Sema] Consider outer max/min selection deprecated.

We warn and provide fixits, so that we're able restrict this behaviour to Swift
4 only by eagerly migrating people off of it.

huonw commented:
I've added the warning/fixits:

```
name_lookup_min_max_conditional_conformance.swift:5:16: warning: use of 'max' as reference to global function in module 'Swift' will change in future versions of Swift to reference instance method in generic struct 'Range' which comes via a conditional conformance
        return max(lowerBound, min(value, upperBound))
               ^
name_lookup_min_max_conditional_conformance.swift:5:16: note: use 'Swift.' to continue to reference the global function
        return max(lowerBound, min(value, upperBound))
               ^
               Swift.
```

huonw commented:
@swift-ci please smoke test.

DougGregor commented:
Warnings/Fix-Its look great, thank you.
